sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/QueryPlanConstraints.scala

```diff
@@ -75,8 +86,10 @@ trait ConstraintHelper {
   private def replaceConstraints(
       constraints: Set[Expression],
       source: Expression,
-      destination: Attribute): Set[Expression] = constraints.map(_ transform {
+      destination: Expression): Set[Expression] = constraints.map(_ transform {
     case e: Expression if e.semanticEquals(source) => destination
+  }).map(_ transform {
+    case Cast(e @ Cast(child, _, _), dt, _) if e == destination && dt == child.dataType => child
```

Comment:
To revert these changes `(cast(cast(int_column#2 as bigint) as int) = 2)` to `(int_column#2 = 2)`.